trie/verkle.go

```diff
@@ -0,0 +1,388 @@
+// Copyright 2021 go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package trie
+
+import (
+	"crypto/sha256"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethdb"
+	"github.com/ethereum/go-ethereum/trie/trienode"
+	"github.com/ethereum/go-ethereum/trie/utils"
+	"github.com/gballet/go-verkle"
+	"github.com/holiman/uint256"
+)
+
+// VerkleTrie is a wrapper around VerkleNode that implements the trie.Trie
+// interface so that Verkle trees can be reused verbatim.
+type VerkleTrie struct {
+	root       verkle.VerkleNode
+	db         *Database
+	pointCache *utils.PointCache
+	ended      bool
+	reader     *trieReader
+}
+
+func (vt *VerkleTrie) ToDot() string {
+	return verkle.ToDot(vt.root)
+}
+
+func NewVerkleTrie(root verkle.VerkleNode, db *Database, pointCache *utils.PointCache, ended bool) (*VerkleTrie, error) {
+	comm := root.Commit().Bytes()
+	rootHash := sha256.Sum256(comm[:])
+	reader, err := newTrieReader(rootHash, common.Hash{}, db)
+	if err != nil {
+		return nil, err
+	}
+	return &VerkleTrie{
+		root:       root,
+		db:         db,
+		pointCache: pointCache,
+		ended:      ended,
+		reader:     reader,
+	}, nil
+}
+
+func (trie *VerkleTrie) FlatdbNodeResolver(path []byte) ([]byte, error) {
+	// NOTE: I use common.Hash{} as the hash, as I expect it to be ignored
+	// since we are using the pathdb. @rjl493456442 please confirm that this
+	// works.
+	return trie.reader.reader.Node(trie.reader.owner, path, common.Hash{})
```

Comment:
@rjl493456442 see question in comment. Because verkle nodes have no reference to their children in path-based mode, it would break the whole scheme if the hash was required in path-based mode.